File: Interview/DP/Kadane.cpp
// Kadane's Algorithm
#include<bits/stdc++.h>
using namespace std;

#define ll long long

int solve(vector<ll> &arr){
	int n = arr.size(); // Array Size

    int csum = INT_MIN, max_sum = 0; 
  
    for (int i = 0; i < n; i++) 
    { 
        max_sum = max_sum + arr[i];

        if (csum < max_sum) 
            csum = max_sum; 
  
        if (max_sum < 0) 
            max_sum = 0; 
    }
    
    return csum; 
}

int main(){
	// ...
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    ll n; cin >> n;
    
    vector<ll> v(n);
    
    for(int i=0;i<n;++i)
    {
        ll x; cin >> x;
        v.push_back(x);
    }
    
    ll ans = solve(v);
    cout << ans;   
}

File: Interview/DP/Kadane.java
// Kadane ALgorithm for Maximum Subarray
import java.io.*;
import java.util.*;

public class Main {
	public static int solution(int[] arr){
		int csum = arr[0];
		int osum = arr[0];

		for(int i = 1; i< arr.length; i++){
			if (csum >= 0) {
				csum += arr[i];
			} else {
				csum = arr[i];
			}

			if (csum > osum) {
				osum = csum;
			}
		}

		return osum;
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		int [] arr = new int[n];

		for (int i = 0; i < arr.length; i++ ) {
			arr[i] = sc.nextInt();
		}
		System.out.println(solution(arr));
	}
}